### DISCUSSION TITLE
questions about eda_process() and eda_intervalrelated()

### DISCUSSION THREAD

**Discussion by alanlflee** [neuropsychology/NeuroKit#870] [Q&A]
Hello,

I'm new to neurokit and would like to ask the following quesitons

1. For `eda_process()`, is the unit of measurement the same for values stored in `"EDA_Clean"`, `"EDA_Phasic"`, and `"EDA_Tonic"` of the output `"signals"` data frame?  I suppose they should be the same as the raw input `eda_signal`, which is in microsiemens in our case.
2.  For `eda_intervalrelated(),` what is the unit of measurement for the `"SCR_Peaks_Amplitude_Mean"` column in the output data frame? I suppose they're in siemens as they're typically numbers at around 1e-7 to 1e-6.
3. For `eda_process()`, I would like to know more about the steps underlying the pre-processing. I'd like to find out about details such as how the data smoothing is done, whether there is any resampling of the data to a lower frequency (if so, how it is done), and whether any of phasic data have been screened out (e.g., those with amplitudes too small/large or rising time toe short/long), and, if no, how we do it manually.

Thank you in advance!

Alan

**Comment by DominiqueMakowski** [comment]
Hi Alan,

1. Once you decompose the raw signal that is in microsiemens the remaining signals lose their unit. Depending on the decomposition method, the phasic is sometimes standardized (mean 0, SD 1).
2. The peak amplitude is calculated on the phasic signal and is relative to that signal's scale. 
3. I'd invite you to take a look at [the code](https://github.com/neuropsychology/NeuroKit/blob/master/neurokit2/eda/eda_process.py#L92) and check all the steps in details.

Hope that helps

**Reply by alanlflee** [reply]
Hi Dom,

Thanks for the quick response. I tried looking in to the code but I can't figure out which decomposition method gives standardized units as output, which one preserves the original unit. Would you mind giving me some hint on this? In fact, I have been using the default methods only, so may I ask if the units are standardized or raw for the default configuration?

Also, just in case it's standardized, may I ask how the mean and SD are computed? is it across all peaks observed within the entire signal period?

Thank you,
alan

**Reply by alanlflee** [reply]
Hi Dom,

I used `eda_intervalrelated()` to obtain some measurements on SCR signals within some predefined epochs. But I don't understand how the values of the `"SCR_Peaks_Amplitude_Mean"` column are computed. I tried extracting the same values directly from the output of `eda_process()`. Not sure if I'm missing something but the values extracted using these two methods are different.

Below is some code I wrote to highlight the difference. You can download the data via [this link](https://drive.google.com/file/d/1gm1XFa3QjWeoVlofP5W_D9VTawyFHJk8/view?usp=sharing).

It'll be great if you could share some insights with us on this!

Best,
Alan

```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import neurokit2 as nk
import pandas as pd
import numpy as np
import scipy as sp

# channel names from each raw .acq file
eda_channel_name = 'EDA1 - GSR100C'
sound_channel_name = 'PO SOUND - Custom, UIM100C - A 2'

# sampling rates
sampling_rate = 2000

# note: all trials have AT LEAST 602,000 (301 seconds) of samples
num_samples_included = 301 * sampling_rate # no. of seconds * sampling_rate

# define epochs
onset_sampleID = [180000, 280000, 380000, 480000]
start_time = [0.0 for o in onset_sampleID]
end_time = [30.0 for o in onset_sampleID] # 30 seconds for each epoch
end_sampleID = []
for o,s,e in zip(onset_sampleID,start_time,end_time):
    end_sampleID.append( o+int((e-s)*sampling_rate) )
print(end_sampleID)

# specify input data file name
datafilename = 'data.acq'

# read the desired .acq file
data, sampling_rate_from_file = nk.read_acqknowledge(datafilename)

# get the EDA data
eda_data = data[eda_channel_name]

# process raw EDA data using eda_process() with default parameters
signals, info = nk.eda_process(eda_data, sampling_rate=sampling_rate)

# create epochs using epochs_create() with default parameters
epochs = nk.epochs_create(signals, onset_sampleID, sampling_rate=sampling_rate,
                          epochs_start=start_time, epochs_end=end_time)

# interval-related analysis using eda_intervalrelated() with default parameters
ir_results = nk.eda_intervalrelated(epochs, sampling_rate=sampling_rate)

# extract SCR information manually

# for each epoch
for epochi, (o, e) in enumerate(zip(onset_sampleID, end_sampleID)):
    
    # get the samples within the epoch
    epoch_signal_id = (signals.index.values >= o) & (signals.index.values < e)
    
    # get the signals falling within this epoch
    epoch_signals = signals.loc[epoch_signal_id]
    
    # extract the peaks falling within this epoch
    peaks_in_epoch = epoch_signals.loc[epoch_signals['SCR_Peaks']==1]
    
    # ====== get mean SCR peak amplitude from two sources =======
    # compute the mean of the amplitudes across the extracted peaks
    mean_peak_amp_from_signals = np.nanmean(peaks_in_epoch['SCR_Amplitude'])
    
    # get the mean peak amplitudes from the output of the interval-related analysis
    mean_peak_amp_from_ir_analysis = ir_results['SCR_Peaks_Amplitude_Mean'][epochi]
    
    # ====== get SD of tonic data as a comparison =======
    # from signals
    sd_tonic_from_signals = np.std(epoch_signals['EDA_Tonic'])
    # from interval-related output
    sd_tonic_from_ir_analysis = ir_results['EDA_Tonic_SD'][epochi]
    
    # print output
    print('\n')
    print('===== Epoch %d =====' % (epochi+1))
    print('Mean SCR peak amplitudes')
    print('  from signals: %g\t From interval-related analysis: %g' % (mean_peak_amp_from_signals, mean_peak_amp_from_ir_analysis))
    print('SD of Tonic')
    print('  from signals: %g\t From interval-related analysis: %g' % (sd_tonic_from_signals, sd_tonic_from_ir_analysis))

print('\n')
```
